Added PAN token to defaults.json, added coingeckoid property, impleme…

arbreton opened the pull request:
…nted coingecko API, for PAN usage, prepared for future usage

aminlatifi commented:
@arbreton 
* The code has eslint problems, fix those please.
* As I understood you fetch the price from today 00:00:00  to the timestamp the rate is requested. What if timestamp is sometime in the middle of today?
* Code walks through the rates to find the closest time to timestamp. We cannot fetch the hourly rates and find the exact match? Can we archive the rates (not just for the timestamp) to use sometimes latter?
* The coinGecko logic is implemented in `_getRatesCryptocompare` function, it's not cool and it should be implemented in separate function with related name.

arbreton commented:
> @arbreton
> 
> * The code has eslint problems, fix those please.
> * As I understood you fetch the price from today 00:00:00  to the timestamp the rate is requested. What if timestamp is sometime in the middle of today?
> * Code walks through the rates to find the closest time to timestamp. We cannot fetch the hourly rates and find the exact match? Can we archive the rates (not just for the timestamp) to use sometimes latter?
> * The coinGecko logic is implemented in `_getRatesCryptocompare` function, it's not cool and it should be implemented in separate function with related name.

Updated the code base, moved coinGecko implementation to a proper function, fixed ESLint problems

Answering the following questions right now seems that we don't have a way to request various timestamps because we only have a date picker on the milestone creation window
Our current base is not enabled for daily rates to my understanding at this point, it will be an interesting feature

aminlatifi commented on src/services/conversionRates/getConversionRatesService.js:
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
+ * @param {Number} timestamp   Timestamp for which the value should be retrieved
+ * @param {Array}  ratestToGet Rates that are missing in the DB and should be retrieved
+ * @param {Array}  stableCoins coins whose value equal one usd
+ *
```
These lines are copied from `_getRatesCryptocompare` function doc and are not correct

arbreton replied:
Fixed this block

aminlatifi commented on src/services/conversionRates/getConversionRatesService.js:
```diff
@@ -47,13 +97,17 @@ const _getRatesCryptocompare = async (timestamp, ratesToGet, symbol, stableCoins
-
-      if (resp && resp[rateSymbol]) rates[r] = resp[rateSymbol];
+      if (requestSymbol === 'PAN') {
+        rates[r] = await _getRatesCoinGecko(rateSymbol, timestampMS, coingeckoId);
```
It's still called inside `_getRatesCryptocompare` and not cool

arbreton replied:
fixed

aminlatifi commented on src/services/conversionRates/getConversionRatesService.js:
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
+
+    const { prices } = testRep;
+
+    for (i = 0; i < prices.length; i += 1) {
```
use forEach

arbreton replied:
fixed

aminlatifi commented on src/services/conversionRates/getConversionRatesService.js:
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
+    const { prices } = testRep;
+
+    for (i = 0; i < prices.length; i += 1) {
+      cur = prices[i];
```
Use array destructuring please

arbreton replied:
How specifically? not sure

arbreton replied:
This i'm not sure yet

aminlatifi commented on src/services/conversionRates/getConversionRatesService.js:
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
+
+    for (i = 0; i < prices.length; i += 1) {
+      cur = prices[i];
+      const curRounded = Math.round(cur[0] / 1000);
```
curRound can be removed

arbreton replied:
Fixed

aminlatifi commented on src/services/conversionRates/getConversionRatesService.js:
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
+  const daysMax = Math.round(differenceDays);
+  const testRep = JSON.parse(
+    await rp(
+      `https://api.coingecko.com/api/v3/coins/${coingeckoId}/market_chart?vs_currency=${rateSymbol}&days=${daysMax}`,
```
Let's use `/coins/{id}/market_chart/range` api for efficiency. Send request for some finite range before and after the requested timestamp, 10 minutes maybe enough.

arbreton replied:
Didn't understood this

arbreton replied:
Oh i understand this one, i will check it out!

aminlatifi commented on src/services/conversionRates/getConversionRatesService.js:
```diff
@@ -148,6 +202,18 @@ const getConversionRates = async (app, requestedDate, requestedSymbol = 'ETH') =
+  const tokens = app.get('activeTokenWhitelist');
+  let coingeckoId = '';
+
+  const keys = Object.keys(tokens);
```
tokens is an array... it's inappropriate? Object.keys or Object.values are used for iterating over objects. Use other standard iterating over array.

arbreton replied:
Not sure what you mean?, will need an example

arbreton replied:
What would be the alternative?

arbreton commented on src/services/conversionRates/getConversionRatesService.js:
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
+ * @return {Object} Rates object in format { EUR: 241, USD: 123 }
+ */
+const _getRatesCoinGecko = async (rateSymbol, timestampMS, coingeckoId) => {
+  const today = new Date(new Date().setUTCDate(new Date().getUTCDate()));
```
Really? I think that's the way it is on line 192, we should change it as well maybe

aminlatifi replied:
Line 192 is different, it has minus one. Also setUTCDate returns number and not date

arbreton commented on src/services/conversionRates/getConversionRatesService.js:
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
+  const today = new Date(new Date().setUTCDate(new Date().getUTCDate()));
+  const todayUTC = today.setUTCHours(0, 0, 0, 0);
+  const todayRounded = Math.round(todayUTC / 1000);
+  const differenceDays = (todayRounded - timestampMS) / 86400;
```
That's interesting and a good observation, i think i have an idea!

aminlatifi replied:
Line 192 is different, it has minus one

arbreton commented:
I think it got merged sooner than expected, if something need to be changed we can definitely update it, so we can release PAN !! wooho!

aminlatifi commented:
Reverted the merge. @arbreton Please fix the problems mentioned in review. If you need any help let me know.
